sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/ChangeFeedFetcher.java

```diff
@@ -0,0 +1,169 @@
+        Supplier<RxDocumentServiceRequest> createRequestFunc,
+        Function<RxDocumentServiceRequest, Mono<FeedResponse<T>>> executeFunc,
+        ChangeFeedState changeFeedState,
+        int top,
+        int maxItemCount) {
+
+        super(executeFunc, true, top, maxItemCount);
+
+        checkNotNull(client, "Argument 'client' must not be null.");
+        checkNotNull(createRequestFunc, "Argument 'createRequestFunc' must not be null.");
+        checkNotNull(changeFeedState, "Argument 'changeFeedState' must not be null.");
+        this.createRequestFunc = createRequestFunc;
+        this.changeFeedState = changeFeedState;
+        this.feedRangeContinuationSplitRetryPolicy = new FeedRangeContinuationSplitRetryPolicy(
+            client,
+            this.changeFeedState);
+    }
+
+    @Override
+    public Mono<FeedResponse<T>> nextPage() {
+        // There are two conditions that require retries
+        // in the change feed pipeline
+        // 1) On 410 the FeedRangeContinuation needs to evaluate
+        //    whether continuations need to be split (in the case that any continuation
+        //    exists that would span more than one physical partition now
+        // 2) On 304 a retry is needed if at least one continuation has not been drained yet.
+        //    This prevents returning a 304 before we received a 304 for all continuations
+        //
+        // 410 handling: this is triggered by an exception - using an
+        //               IRetryPolicy (FeedRangeContinuationSplitRetryPolicy)
+        // 304 handling: this is not triggered by an exception (304 doesn't result in throwing)
+        //               so using Reactor's built-in option of repeating the chain on an empty result
+        //               so nextPageInternal below has the logic to return empty result
+        //               if not all continuations have been drained yet.
+        return ObservableHelper.inlineIfPossible(
+            this::nextPageInternal,
+            this.feedRangeContinuationSplitRetryPolicy);
+    }
+
+    private Mono<FeedResponse<T>> nextPageInternal() {
+        return Mono.fromSupplier(this::nextPageCore)
+                   .flatMap(Function.identity())
+                   .flatMap((r) -> {
+                       FeedRangeContinuation continuationSnapshot =
+                           this.changeFeedState.getContinuation();
+
+                       if (continuationSnapshot != null &&
+                           continuationSnapshot.handleChangeFeedNotModified(r) == ShouldRetryResult.RETRY_NOW) {
+
+                           // not all continuations have been drained yet
+                           // repeat with the next continuation
+                           this.reenableShouldFetchMoreForRetry();
+                           return Mono.empty();
+                       }
+
+                       return Mono.just(r);
+                   })
+                   .repeatWhenEmpty(o -> o);
```

Comment:
Good point - it is a very subtle contract - basically because handleChangeFeedNotModified is guaranteed to not always return Retry_Now because it moves ahead the continuation token. Let me think about a way to cover this better in unit tests - or how to change the contract to be more explicit - will address this in subsequent PR